events/janus_gelfevh.c

```diff
@@ -0,0 +1,589 @@
+static char *backend = NULL;
+static char *port = NULL;
+
+typedef enum janus_gelfevh_socket_type {
+	TCP = 1,
+	UDP = 2
+} janus_gelfevh_socket_type;
+
+static int max_gelf_msg_len = 500;
+static int sockfd;
+/* Set TCP as Default transport */
+static janus_gelfevh_socket_type transport = TCP;
+
+/* Parameter validation (for tweaking via Admin API) */
+static struct janus_json_parameter request_parameters[] = {
+	{"request", JSON_STRING, JANUS_JSON_PARAM_REQUIRED}};
+static struct janus_json_parameter tweak_parameters[] = {
+	{"events", JSON_STRING, 0},
+	{"backend", JSON_STRING, 0},
+	{"port", JSON_STRING, 0},
+	{"max_gelf_msg_len", JSON_INTEGER, JANUS_JSON_PARAM_POSITIVE},
+	{"janus_gelfevh_socket_type", JSON_STRING, 0}
+};
+/* Error codes (for the tweaking via Admin API */
+#define JANUS_GELFEVH_ERROR_INVALID_REQUEST		411
+#define JANUS_GELFEVH_ERROR_MISSING_ELEMENT		412
+#define JANUS_GELFEVH_ERROR_INVALID_ELEMENT		413
+#define JANUS_GELFEVH_ERROR_UNKNOWN_ERROR		499
+
+/* Plugin implementation */
+static char *randstring(size_t length) {
+	static char charset[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+	char *randomString = NULL;
+	if (length) {
+		randomString = malloc(sizeof(char) * (length + 1));
+		if (randomString) {
+			for (int n = 0; n < (int)length; n++) {
+				int key = rand() % (int)(sizeof(charset) - 1);
+				randomString[n] = charset[key];
+			}
+			randomString[length] = '\0';
+		}
+	}
+	return randomString;
+}
+
+static int janus_gelfevh_connect(void) {
+	struct addrinfo *res = NULL;
+	janus_network_address addr;
+	janus_network_address_string_buffer addr_buf;
+	struct sockaddr_in servaddr;
+
+	if(getaddrinfo(backend, NULL, NULL, &res) != 0 ||
+				janus_network_address_from_sockaddr(res->ai_addr, &addr) != 0 ||
+				janus_network_address_to_string_buffer(&addr, &addr_buf) != 0) {
+		if(res)
+			freeaddrinfo(res);
+		JANUS_LOG(LOG_ERR, "Could not resolve address (%s)...\n", backend);
+		return -1;
+	}
+	const char *host = g_strdup(janus_network_address_string_from_buffer(&addr_buf));
+	freeaddrinfo(res);
+
+    if((sockfd = socket(AF_INET, transport, 0)) < 0 ) {
+		JANUS_LOG(LOG_ERR, "Socket creation failed: %s\n", strerror(errno));
+		return -1;
+	}
+
+	memset(&servaddr, 0, sizeof(servaddr));
+
+	servaddr.sin_family = AF_INET;
+	servaddr.sin_port = htons(atoi(port));
+	servaddr.sin_addr.s_addr = inet_addr(host);
+
+	if(connect(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
+		JANUS_LOG(LOG_ERR, "Connect to GELF host failed\n");
+		return -1;
+	}
+	JANUS_LOG(LOG_INFO, "Connected to GELF backend: [%s:%s]\n", host, port);
+	return 1;
+}
+
+static int janus_gelfevh_send(char *message) {
+	if(!message) {
+		JANUS_LOG(LOG_WARN, "Message is NULL, not sending to GELF!\n");
+		return -1;
+	}
+	if(transport == TCP) {
+		/* TCP */
+		unsigned int out_bytes = 0;
+		while (out_bytes < strlen(message)+1) {
+			int n = send(sockfd, message, strlen(message) + 1, 0);
```

Comment:
I think this is broken. You're looping until a number of bytes equal to the size of the string is sent, but then in case you only sent part of them, you're sending the whole string again. This should try to send the part of the string that wasn't sent (using an offset, and the updated length). As a side note, no need for the +1, you typically don't need to send the terminator.

Reply:
I think it should be ok like this